sklearn/linear_model/_logistic.py

```diff
@@ -397,34 +397,54 @@ def _logistic_regression_path(
             # reconstructs the 2d-array via w0.reshape((n_classes, -1), order="F").
             # As w0 is F-contiguous, ravel(order="F") also avoids a copy.
             w0 = w0.ravel(order="F")
-            loss = LinearModelLoss(
+            linear_loss = LinearModelLoss(
                 base_loss=HalfMultinomialLoss(n_classes=classes.size),
                 fit_intercept=fit_intercept,
             )
         target = Y_multi
-        if solver in "lbfgs":
-            func = loss.loss_gradient
-        elif solver == "newton-cg":
-            func = loss.loss
-            grad = loss.gradient
-            hess = loss.gradient_hessian_product  # hess = [gradient, hessp]
         warm_start_sag = {"coef": w0.T}
     else:
         target = y_bin
         if solver == "lbfgs":
-            loss = LinearModelLoss(
+            linear_loss = LinearModelLoss(
                 base_loss=HalfBinomialLoss(), fit_intercept=fit_intercept
             )
-            func = loss.loss_gradient
         elif solver == "newton-cg":
-            loss = LinearModelLoss(
+            linear_loss = LinearModelLoss(
                 base_loss=HalfBinomialLoss(), fit_intercept=fit_intercept
             )
-            func = loss.loss
-            grad = loss.gradient
-            hess = loss.gradient_hessian_product  # hess = [gradient, hessp]
         warm_start_sag = {"coef": np.expand_dims(w0, axis=1)}
 
+    if solver == "lbfgs":
+        func = linear_loss.loss_gradient
```

Comment:
Can we use the following to avoid using `temporary_array_dict`?

```
func = functools.partial(func,
    per_sample_loss_out=per_sample_loss_out, 
    per_sample_gradient_out=per_sample_gradient_out)
```

This also means we do not need to pass `per_sample_*_out` into `args` belong.

This way `linear_loss.gradient` can accept `per_sample_*_out` directly, without needing to go through the logic of parsing the `temporary_array_dict`.